services/src/main/java/org/keycloak/protocol/saml/SamlMetadataKeyLocator.java

```diff
@@ -0,0 +1,123 @@
+ *
+ * @author rmartinc
+ */
+public class SamlMetadataKeyLocator implements KeyLocator {
+
+    private final String modelKey;
+    private final PublicKeyLoader loader;
+    private final PublicKeyStorageProvider keyStorage;
+    private final KeyUse use;
+
+    public SamlMetadataKeyLocator(String modelKey, PublicKeyLoader loader, KeyUse use, PublicKeyStorageProvider keyStorage) {
+        this.modelKey = modelKey;
+        this.loader = loader;
+        this.keyStorage = keyStorage;
+        this.use = use;
+    }
+
+    @Override
+    public Key getKey(String kid) throws KeyManagementException {
+        if (kid == null) {
+            return null;
+        }
+        // search the key by kid and reload if expired or null
+        KeyWrapper keyWrapper = keyStorage.getPublicKey(modelKey, sameKidPredicate(kid), loader);
+        return keyWrapper != null? keyWrapper.getPublicKey() : null;
+    }
+
+    @Override
+    public Key getKey(Key key) throws KeyManagementException {
+        if (key == null) {
+            return null;
+        }
+        // search the key and reload if expired or null
+        KeyWrapper keyWrapper = keyStorage.getPublicKey(modelKey, sameKeyPredicate(key), loader);
+        return keyWrapper != null? keyWrapper.getPublicKey() : null;
+    }
+
+    @Override
+    public void refreshKeyCache() {
+        keyStorage.reloadKeys(modelKey, loader);
+    }
+
+    @Override
+    public Iterator<Key> iterator() {
+        // force a refresh if a certificate is expired?
+        return keyStorage.getKeys(modelKey, loader)
+                .stream()
+                .filter(k -> use.equals(k.getUse()) && isValidCertificate(k))
+                .map(KeyWrapper::getPublicKey)
+                .iterator();
+    }
+
+    private Predicate<KeyWrapper> sameKidPredicate(String kid) {
+        return keyWrapper -> isSameKid(keyWrapper, kid);
+    }
+
+    private boolean isSameKid(KeyWrapper keyWrapper, String kid) {
+        String k = keyWrapper.getKid();
+        if (k == null) {
+            return false;
+        }
+        return k.equals(kid) && use.equals(keyWrapper.getUse()) && isValidCertificate(keyWrapper);
+    }
+
+    private Predicate<KeyWrapper> sameKeyPredicate(Key key) {
+        return keyWrapper -> isSameKey(keyWrapper, key);
+    }
+
+    private boolean isSameKey(KeyWrapper keyWrapper, Key key) {
+        Key k = keyWrapper.getPublicKey();
+        if (k == null) {
+            return false;
+        }
+        return use.equals(keyWrapper.getUse())
+                && key.getAlgorithm().equals(k.getAlgorithm())
+                && MessageDigest.isEqual(k.getEncoded(), key.getEncoded())
+                && isValidCertificate(keyWrapper);
+    }
+
+    private boolean isValidCertificate(KeyWrapper key) {
+        if (key == null || key.getCertificate() == null) {
+            return false;
+        }
+        try {
+            key.getCertificate().checkValidity();
```

Comment:
Doesn't make more sense to actually enter in an error state if the certificate is not valid? What about other validations like checking if the certificate can be trusted by the configured trust store?

Reply:
Imagine a situation that they are caching certificates for one week for example. The certificate expires in that week but the IdP correctly replaced it before expiration. This check will ignore the expired one, there will be a reload because no key was found, and the new certificate (not expired) will be loaded and returned. As it's just a cache I think we should consider that certificates can expire in it, they are just not returned, triggering as expected.

Reply:
I see now, makes sense for me. As we are operating in the cache we also don't want to validate it against the trust store.